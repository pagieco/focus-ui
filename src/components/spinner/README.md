# Spinner

Spinners are used to notify users that their action is being processed. For loading states, spinners should only be used
for content that can't be represented with a skeleton loading component, like for data charts.

---

## Examples

---

## Best practices

The spinner component should:

- Notify users that their request has ben received and the action will soon complete
- Not be used to given feedback for an entire page load
- White can only be used with small spinners on actionable components like buttons
- Be used in conjunction with skeleton loading to represent non-typographic content. For example, line graphs

---

## Content guidelines

### Accessibility label

Spinner accessibility label should:

- Accurately explaint he state of the requested action. For example, "Loading", "Submitting", "Processing".
- Use as few words to describe the state as possible.
