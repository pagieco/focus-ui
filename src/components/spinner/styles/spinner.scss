.spinner {
  display: inline-flex;
  animation: spinner-animation 0.5s linear infinite;

  svg {
    display: block;
  }

  path {
    fill-opacity: 0;
  }

  .spinner__track {
    stroke: rgba(0, 0, 0, 0.15);
  }

  .spinner__head {
    stroke: white;
    stroke-linecap: round;
    transition: stroke-dashoffset 0.2s cubic-bezier(0.4, 1, 0.75, 0.9);
    transform-origin: center;
  }

  @keyframes spinner-animation {
    0% { transform: rotate(0deg); }
    to { transform: rotate(1turn); }
  }
}
