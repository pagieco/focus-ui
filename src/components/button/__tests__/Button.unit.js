import { shallowMount } from '@vue/test-utils';
import { Spinner } from '../../spinner';
import { Button } from '../index';

const shallowWrap = propsData => shallowMount(Button, { propsData });

describe('Button', () => {
  it('is a valid vue component', () => {
    expect(Button).toBeAComponent();
  });

  it('matches the snapshot', () => {
    expect(shallowMount(Button).element).toMatchSnapshot();
  });

  it('uses the ariaLabel prop', () => {
    const wrapper = shallowWrap({ ariaLabel: 'label' });

    expect(wrapper.attributes('aria-label')).toBe('label');
  });

  it('uses the ariaControls prop', () => {
    const wrapper = shallowWrap({ ariaControls: 'label' });

    expect(wrapper.attributes('aria-controls')).toBe('label');
  });

  it('uses the ariaExpanded prop', () => {
    const wrapper = shallowWrap({ ariaExpanded: true });

    expect(wrapper.attributes('aria-expanded')).toBe('true');
  });

  it('uses the id prop', () => {
    const wrapper = shallowWrap({ id: 'my-component-id' });

    expect(wrapper.attributes('id')).toBe('my-component-id');
  });

  it('uses the disabled prop', () => {
    const wrapper = shallowWrap({ disabled: true });

    expect(wrapper.classes()).toContain('button--disabled');
    expect(wrapper.attributes('disabled')).toBe('disabled');
    expect(wrapper.contains(Spinner)).toBeFalsy();
  });

  it('uses the external prop', () => {
    const wrapper = shallowWrap({ external: true, url: 'https://my-amazing.domain' });

    expect(wrapper.attributes('target')).toBe('_blank');
  });

  it('wont use the external prop when the button isnt an anchor', () => {
    const wrapper = shallowWrap({ external: true });

    expect(wrapper.attributes('target')).toBeUndefined();
  });

  it('uses the loading prop', () => {
    const wrapper = shallowWrap({ loading: true });

    expect(wrapper.classes()).toContain('button--disabled');
    expect(wrapper.attributes('disabled')).toBe('disabled');
    expect(wrapper.contains(Spinner)).toBeTruthy();
  });

  it('uses the primary prop', () => {
    const wrapper = shallowWrap({ primary: true });

    expect(wrapper.classes()).toContain('button--primary');
  });

  it('uses the danger prop', () => {
    const wrapper = shallowWrap({ danger: true });

    expect(wrapper.classes()).toContain('button--danger');
  });

  it('uses the minimal prop', () => {
    const wrapper = shallowWrap({ minimal: true });

    expect(wrapper.classes()).toContain('button--minimal');
  });

  it('uses the small prop', () => {
    const wrapper = shallowWrap({ small: true });

    expect(wrapper.classes()).toContain('button--small');
  });

  it('uses the submit prop', () => {
    const wrapper = shallowWrap({ submit: true });

    expect(wrapper.attributes('type')).toBe('submit');
  });

  it('uses the url prop', () => {
    const url = 'https://my-amazing.domain';
    const wrapper = shallowWrap({ url });

    expect(wrapper.attributes('href')).toBe(url);
    expect(wrapper.html().charAt(1)).toBe('a');
  });
});
