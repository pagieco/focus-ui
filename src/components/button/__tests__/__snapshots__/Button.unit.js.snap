// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches the snapshot 1`] = `
<button
  class="button"
  role="button"
  tabindex="0"
  type="button"
>
  <span
    class="button__label"
  />
</button>
`;
